brillig_bytecode/src/lib.rs

```diff
@@ -4,111 +4,126 @@
+pub use opcodes::{Comparison, Opcode};
 pub use registers::{RegisterIndex, Registers};
+use serde::{Deserialize, Serialize};
 pub use value::Typ;
 pub use value::Value;
 
-#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[derive(Debug, PartialEq, Eq, Clone)]
 pub enum VMStatus {
-    Halted,
+    Finished,
     InProgress,
     Failure,
-    OracleWait,
+    ForeignCallWait {
+        // Interpreted by simulator context
+        function: String,
+        // Input values
+        inputs: Vec<Value>,
+    },
 }
 
-#[derive(Default, Debug, PartialEq, Eq, Clone)]
-pub struct ArrayHeap {
-    // maps memory address to Value
-    pub memory_map: BTreeMap<usize, Value>,
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone)]
+pub struct ForeignCallResult {
+    // resolved foreign call values
+    pub values: Vec<Value>,
 }
 
 #[derive(Debug, PartialEq, Eq, Clone)]
 pub struct VM {
     registers: Registers,
     program_counter: usize,
+    foreign_call_counter: usize,
+    foreign_call_results: Vec<ForeignCallResult>,
     bytecode: Vec<Opcode>,
     status: VMStatus,
-    memory: BTreeMap<Value, ArrayHeap>,
+    memory: Vec<Value>,
     call_stack: Vec<Value>,
 }
 
 impl VM {
     pub fn new(
         mut inputs: Registers,
-        memory: BTreeMap<Value, ArrayHeap>,
-        mut bytecode: Vec<Opcode>,
+        memory: Vec<Value>,
+        bytecode: Vec<Opcode>,
+        foreign_call_results: Vec<ForeignCallResult>,
+        register_allocation_indices: Option<Vec<u32>>,
     ) -> VM {
-        let last_opcode = bytecode.last().expect("need at least one opcode");
-
-        if let Opcode::Bootstrap { register_allocation_indices } = last_opcode {
-            // TODO: might have to handle arrays in bootstrap to be correct
-            let mut registers_modified =
-                Registers::load(vec![Value { typ: Typ::Field, inner: FieldElement::from(0u128) }]);
+        if let Some(register_allocation_indices) = register_allocation_indices {
+            // TODO(AD): simplify this all to be done before calling VM.new()
+            let mut registers_modified = Registers::load(vec![Value::from(0u128)]);
 
-            for i in 0..register_allocation_indices.len() {
-                let register_index = register_allocation_indices[i];
-                let register_value = inputs.get(RegisterMemIndex::Register(RegisterIndex(i)));
-                registers_modified.set(RegisterIndex(register_index as usize), register_value)
+            for (i, register_index) in register_allocation_indices.iter().enumerate() {
+                let register_value = inputs.get(RegisterIndex(i));
+                registers_modified.set(RegisterIndex(*register_index as usize), register_value)
             }
 
-            bytecode.pop();
             inputs = registers_modified;
         }
-        let vm = Self {
+
+        Self {
             registers: inputs,
             program_counter: 0,
+            foreign_call_counter: 0,
```

Comment:
Whenever we construct a VM we start the foreign_call_counter at 0. For someone using the brillig VM from acvm a new VM is currently constructed in the brillig solver, meaning we will restart at `0` each time the ACVM stalls. It seems like we will need to pass in counter information. We should make a test in `acvm/src/lib.rs` that performs more than one foreign call. I can take a crack at this a little later.

A potential alternative could be that whenever we run into an `Opcode::ForeignCall` and `foreign_call_results` is empty to resolving the foreign call inputs. Otherwise, we pop the most result foreign call result and handle it appropriately.

Reply:
Actually we re-run from the start in brillig sovler every time so this is ok. But still good to add a test with multiple foreign calls

Reply:
Added a test in commit: 6ab600ccd3250e96701567cabcdeb11a9c594820